File: SDK Examples/Examples/VB.Net/Counter_PulseWidthMeasurement/PulseWidthMeasurementForm.vb
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports System.Collections
Imports System.Drawing.Drawing2D
Imports Automation.BDaq

Partial Public Class PulseWidthMeasurementForm
	Inherits Form
	Private m_ECisRunning As Boolean = True
	Private m_isPwMeterReset As Boolean = True
	Const m_maxOfCounterValueList As Integer = 7
	' The max value of lines of CounterValueList.
	Public Sub New()
		InitializeComponent()
	End Sub

	Private Sub PulseWidthMeasurementrForm_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
		'no device is selected
		If Not m_PwMeterCtrl.Initialized Then
			MessageBox.Show("Please select a device in control property!")
			Me.Close()
			Return
		End If
		Dim t As Integer = m_PwMeterCtrl.Features.ChannelCountMax
		'set title of the form.
		Dim description As String = m_PwMeterCtrl.SelectedDevice.Description
		Me.Text = "Pulse Width Measurement - Run(" & description & ")"

		Timer1.Enabled = False
		Stop_button.Enabled = False
	End Sub

	Private Sub timer1_Tick(ByVal sender As Object, ByVal e As EventArgs) Handles Timer1.Tick
		Dim value As PulseWidth

		Try
			If m_ECisRunning Then
				' Show event counting value
				m_PwMeterCtrl.Read(value)
				CurHiPeriod.Text = value.HiPeriod.ToString("F5")
				CurLoPeriod.Text = value.LoPeriod.ToString("F5")
			End If
		Catch ex As Exception
			ShowErrorMessage(ex)
		End Try
	End Sub

	Private Sub Start_Click(ByVal sender As Object, ByVal e As EventArgs) Handles Start_button.Click
		If m_isPwMeterReset = True Then
			CounterValueList.Items.Clear()
			Try
				m_PwMeterCtrl.Enabled = True
				m_ECisRunning = True
				If False = Timer1.Enabled Then
					Timer1.Start()
				End If
				Stop_button.Enabled = True
				Start_button.Text = "Latch"
				m_isPwMeterReset = False
			Catch ex As Exception
				ShowErrorMessage(ex)
			End Try
		Else
			CounterValueList.Items.Add("High:" + CurHiPeriod.Text + " Low:" + CurLoPeriod.Text)
			If CounterValueList.Items.Count > m_maxOfCounterValueList Then
				CounterValueList.Items.RemoveAt(0)
			End If
		End If
	End Sub

	Private Sub Stop_Click(ByVal sender As Object, ByVal e As EventArgs) Handles Stop_button.Click
		' Stop event counting
		Try
			m_PwMeterCtrl.Enabled = False
			m_ECisRunning = False
		Catch ex As Exception
			ShowErrorMessage(ex)
		End Try

		m_isPwMeterReset = True
		Start_button.Enabled = True
		Start_button.Text = "Start"
		Stop_button.Enabled = False
		'stop the timer
		If False = m_ECisRunning Then
			Timer1.Enabled = False
		End If
	End Sub
	Private Shared Sub ShowErrorMessage(ByVal e As Exception)
		Dim errorInfo As String
		errorInfo = "There's some error happened, the error information: "
		MessageBox.Show(errorInfo & e.Message)
	End Sub
End Class

File: SDK Examples/Examples/C#/Counter_PulseWidthMeasurement/PulseWidthMeasurementForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using System.Collections;
using System.Drawing.Drawing2D;
using Automation.BDaq;


namespace Counter_PulseWidthMeasurement
{
   public partial class PulseWidthMeasurementForm : Form
   {
      bool m_ECisRunning = true;
      bool m_isPwMeterReset = true;
      const int m_maxOfCounterValueList = 7; // The max value of lines of CounterValueList.

      public PulseWidthMeasurementForm()
      {
         InitializeComponent();
      }

      private void PulseWidthMeasurementForm_Load(object sender, EventArgs e)
      {
         //The default device of project is demo device, users can choose other devices according to their needs. 
         //// no device is selected
         if (!m_pwMeterCtrl.Initialized)
         {
            MessageBox.Show("No device be selected or device open failed!");
            this.Close();
            return;
         }
         int t = m_pwMeterCtrl.Features.ChannelCountMax;
         //set title of the form.
         string description = m_pwMeterCtrl.SelectedDevice.Description;
         this.Text = "Pulse Width Measurement(" + description + ")";

         timer1.Enabled = false;
         Stop.Enabled = false;
      }

      private void timer1_Tick(object sender, EventArgs e)
      {
			PulseWidth value;
			try
         {
            if (m_ECisRunning)
            {
               // Show PwmIn value
					m_pwMeterCtrl.Read(out value);
					curHiPeriod.Text = value.HiPeriod.ToString("0.00000");
					curLoPeriod.Text = value.LoPeriod.ToString("0.00000");
            }
         }
         catch (Exception ex)
         {
            ShowErrorMessage(ex);
         }
      }

      private void Start_Click(object sender, EventArgs e)
      {
         if (m_isPwMeterReset == true)
         {
            CounterValueList.Items.Clear();
            try
            {
               m_pwMeterCtrl.Enabled = true;
               m_ECisRunning = true;
               if (false == timer1.Enabled)
               {
                  timer1.Start();
               }
               Stop.Enabled = true;
               Start.Text = "Latch";
               m_isPwMeterReset = false;
            }
            catch (Exception ex)
            {
               ShowErrorMessage(ex);
            }
         }
         else
         {
            CounterValueList.Items.Add("High:"+curHiPeriod.Text+" Low:"+curLoPeriod.Text);
            if (CounterValueList.Items.Count > m_maxOfCounterValueList)
            {
               CounterValueList.Items.RemoveAt(0);
            }
         }
      }

      private void Stop_Click(object sender, EventArgs e)
      {
         // Stop event counting
         try
         {
            m_pwMeterCtrl.Enabled = false;
            m_ECisRunning = false;
         }
         catch (Exception ex)
         {
            ShowErrorMessage(ex);
         }

         m_isPwMeterReset = true;
         Start.Enabled = true;
         Start.Text = "Start";
         Stop.Enabled = false;
         //stop the timer
         if (false == m_ECisRunning)
         {
            timer1.Enabled = false;
         }
      }
      private static void ShowErrorMessage(Exception e)
      {
         string errorInfo;
         errorInfo = "There's some error happened, the error information: ";
         MessageBox.Show(errorInfo + e.Message);
      }
   }
}